### PR TITLE
Deprecate Add LoadPath

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1438,8 +1438,12 @@ let vernac_set_end_tac ~pstate tac =
 let expand filename =
   Envars.expand_path_macros ~warn:(fun x -> Feedback.msg_warning (str x)) filename
 
+let warn_add_loadpath = CWarnings.create ~name:"add-loadpath-deprecated" ~category:"deprecated"
+    (fun () -> strbrk "Commands \"Add LoadPath\" and \"Add Rec LoadPath\" are deprecated. Use command-line \"-Q\" or \"-R\" or put them in your _CoqProject instead.")
```
Comment (Alizter) [inline vernac/vernacentries.ml:1442]:
```suggestion
    (fun () -> strbrk "Commands \"Add LoadPath\" and \"Add Rec LoadPath\" are deprecated. Use command-line \"-Q\" or \"-R\" or put them in your _CoqProject file instead.")
```

**Review by Alizter — APPROVED** [review]

**Comment by gares** [comment]
I think you should add `.. deprecated:: 8.16` somewhere in doc/spynx

File: doc/changelog/07-vernac-commands-and-options/15652-depr-add-loadpath.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+- **Deprecated:**
+  :cmd:`Add LoadPath` and :cmd:`Add Rec LoadPath`
```
Comment (Alizter) [inline doc/changelog/07-vernac-commands-and-options/15652-depr-add-loadpath.rst:2]:
```suggestion
  :cmd:`Add LoadPath` and :cmd:`Add Rec LoadPath` (use \"-Q\" or \"-R\" instead)
```

Comment (Alizter) [reply doc/changelog/07-vernac-commands-and-options/15652-depr-add-loadpath.rst:2]: That might be too ambiguous in the changelog however. Maybe also mention command-line/_CoqProject.

**Comment by silene** [comment]
Did they stop using it in ProofGeneral? (Because I think it is the reason it was still around.)

**Comment by gares** [comment]
CC @Matafou @erikmd

**Comment by Alizter** [comment]
There is one in the tests: https://github.com/ProofGeneral/PG/search?p=1&q=Add+LoadPath

**Comment by jfehrle** [comment]
Using `Add LoadPath` in the middle of script clearly seems a poor practice.  On the other hand, I think users should be able to configure the libraries they need without restarting their IDE and perhaps also without restarting coqidetop.

It seems to me we haven't made it easy to configure libraries.  Adding them as options on the command line is dreadful.  The `_CoqProject` mechanism helps, but it's not compositional, e.g. if you need VST and its dependencies, you have to copy all of them into your _CoqProject rather than just pointing to a VST-provided config file (another _CoqProject file??).  Maybe an improved mechanism could automatically include information from opam?

I would consider dropping the `-Q` and `-I` flags in favor of always (implicitly) including with `-R`.  I don't see a good use case for them.  Of course, we'd need to consider how to migrate existing projects.

I'm not so sure what I'd do about `coq_makefile`.  Some (many? most?) projects provide their own makefiles, so _CoqProject isn't a useful way to pass the list of libraries into the makefiles.  Also AFAIK it's irrelevant to building with dune.

Also, the documentation on configuring libraries and building projects needs considerable work.

I wonder if some have continuted to use `Add LoadPath` because it's a pain to use the other mechanisms.

Disclaimer: I've not used these mechanisms much--feel free to disagree/correct me if I've said anything foolish.

We should make it much easier to configure libraries.

**Comment by Matafou** [comment]
> There is one in the tests: https://github.com/ProofGeneral/PG/search?p=1&q=Add+LoadPath

Yes but the file is called `add-load-path-unsupported/README` :-).

**Comment by gares** [comment]
> 8 users who say it would be impossible to do without it, and 7 who say it would be hard to do without it.

Did by any chance write down why? It is hard to provide an alternative if we don't know the use case.

> I'm not so sure what I'd do about `coq_makefile`. Some (many? most?) projects provide their own makefiles

90% of the projects we run Coq's CI on build using coq_makefile. 50% also provide a dune file.

> I wonder if some have continuted to use Add LoadPath because it's a pain to use the other mechanisms.

Writing `-R foo bar` in `_CoqProject` is not more painful that wirting `Add Rec LoadPath foo bar` in *all* yor .v files (all UIs read _CoqProject). IMO these 8 users have some very weird use case, or don't know what I just wrote. (note that even if you build your project with a custom script, you can still have a _CoqProject around to set options).

**Comment by SkySkimmer** [comment]
We could also keep it as an interactive-only command.

**Comment by gares** [comment]
You mean as in `coqtop` but not `coqc`? In this case, I've no clue if this would cover their usecase...

**Comment by Zimmi48** [comment]
> Did by any chance write down why? It is hard to provide an alternative if we don't know the use case.

We didn't provide a field for doing that unfortunately. But what about doing any of my two suggestions (writing on various Coq channels to get more info from these users, or making the deprecation message request these users to come open issues)?

**Comment by gares** [comment]
Im happy with the latter option (and give a longer deprecation time)

**Comment by silene** [comment]
I wonder if people still using `Add LoadPath` are people who learned Coq at a time when it was impossible to install Coq libraries, and thus they had to manually handle the logical paths. In that situation, it was better to use `Add LoadPath` than `-R`, for documentation and reproducibility purpose. This time is, fortunately, long gone.

File: vernac/vernacentries.ml [hunk 3]
```diff
@@ -1438,8 +1438,12 @@ let vernac_set_end_tac ~pstate tac =
 let expand filename =
   Envars.expand_path_macros ~warn:(fun x -> Feedback.msg_warning (str x)) filename
 
+let warn_add_loadpath = CWarnings.create ~name:"add-loadpath-deprecated" ~category:"deprecated"
+    (fun () -> strbrk "Commands \"Add LoadPath\" and \"Add Rec LoadPath\" are deprecated. Use command-line \"-Q\" or \"-R\" or put them in your _CoqProject file instead.")
+
```
Comment (gares) [inline vernac/vernacentries.ml:1443]:
```suggestion
    (fun () -> strbrk "Commands \"Add LoadPath\" and \"Add Rec LoadPath\" are deprecated." ++ spc () ++
               strbrk "Use command-line \"-Q\" or \"-R\" or put them in your _CoqProject file instead." ++ spc () ++
               strbrk "If \"Add [Rec] LoadPath\" is an important feature for you, please open an issue at" ++ spc () ++
               strbrk "https://github.com/coq/coq/issues" ++ spc () ++ strbrk "and explain your workflow.")

```

**Comment by silene** [comment]
> we may consider forcing people to only use `-Q` and explicitly write `From MyProject Require foo` even for modules declared in the local dev.

That is just a way to aggravate users. Since we still need to support `-R` internally for the standard library, it does not cost us anything to expose it to users.

> whether this flag is nowadays useless or if it has still some worthy use case.

Until the compiled plugin is actually installed in a place where Findlib can find it, option `-I` still serves a critical purpose.

**Comment by gares** [comment]
> I haven't followed the findlib patch enough to understand whether `-I` flag is nowadays useless or if it has still some worthy use case.

In the end, the META files are looked up *also* in the paths passed via `-I`, which typically includes `src/` that is where .ml code is (and also where the META is automatically generated by coq_makefile, is applicable). Of course OCAMLPATH is also supported and as well the paths in the configuration of findlib (eg /usr/lib).

In any case it is still used by `ocamlc` to find .cm* files.

**Comment by gares** [comment]
> The backlash from users was impressive, especially from the ones who teach Coq.

I recall one vocal user/dev (Pierre Letouzey). I don't recall if we also asked this question in the previous poll. In any case I find the current situation OK.

**Comment by gares** [comment]
> Disclaimer: I've not used these mechanisms much--feel free to disagree/correct me if I've said anything foolish.

I think you should see this mechanism akin to `JAVA_HOME` or -classpath. I don't know if it is possible, but I'm sure it would not be a nice practice to alter its value from within a `.java` file.

**Comment by jfehrle** [comment]
> > I would consider dropping the -Q and -I flags in favor of always (implicitly) including with -R. I don't see a good use case for them. Of course, we'd need to consider how to migrate existing projects.

> @ppedrot That's literally the opposite, `-R` is the highly non-modular flag and we may consider forcing people to only use `-Q` and explicitly write `From MyProject Require foo` even for modules declared in the local dev.

Maybe so.  My recollection was that one of the flags includes all subdirectories recursively (I thought that was -R) and the other (-Q) doesn't.  The descriptions in the [documentation](https://coq.inria.fr/distrib/current/refman/practical-tools/coq-commands.html#by-command-line-options) are not at all clear.

> As for `-I` it is unrelated to the `-Q / -R` flags. I haven't followed the findlib patch enough to understand whether this flag is nowadays useless or if it has still some worthy use case. 

I find I need `-I` to work with fiat_crypto in CoqIDE, e.g. `dune exec -- dev/shim/coqide-prelude -R _build_ci/fiat_crypto/src Crypto -R _build_ci/rewriter/src/Rewriter Rewriter -I _build_ci/rewriter/src/Rewriter/Util/plugins`

**Comment by jfehrle** [comment]
> Don't use coqide-prelude to work with stuff in _build_ci.

Why not?  What do you suggest as an alternative?
Also, how the heck would I know that?

**Comment by SkySkimmer** [comment]
>Why not? What do you suggest as an alternative?

Normal coqide. The shims are for hacking with a partial build.

**Comment by jfehrle** [comment]
That doesn't make sense, nor have I seen it mentioned anywhere, but let's not dive into it now.

**Comment by jfehrle** [comment]
To improve the configuration process, we should first come up with a list of use cases/requirements.  I think there are more cases than one might expect.

Configuration is one of the first things that users have to deal with.  Often it blocks them from further progress.  Yet it's (hopefully) something that users do very infrequently, so they don't get very good at it.  We should make configuration as simple as possible for users and very well documented.  This may require significant development work.

**Comment by gares** [comment]
FTR maybe Coqide UI is a bit rough, but lets you pass options to coqidetop, among which -R, and if a project has  _CoqProject file it picks options up from there. We all like this.

This PR is about deprecating the vernac command that you can put in the middle of a .v file, and that has a similar effect than -R (but with some intronsic problems).

**Comment by gares** [comment]
The 10% has needs that are not covered by coq makefile, or authors which are fluent in Make. IMO this is fine.

Dune is new, and less flexible than coq_makefile, hence the lower rate. It is more promising though.

**Comment by gares** [comment]
TTBOMK fiat uses coq makefile, wrapped in some custom make logic to support many coq versions.

**Comment by jfehrle** [comment]
One more off-topic thought: IIUC a dirpath can have multiple directories associated with it, which I think is rarely used.  It would be simpler for users and developers to have only one directory for a dirpath--if feasible.

**Review by ejgallego — APPROVED** [review]
LGTM

**Comment by Alizter** [comment]
@SkySkimmer It's not obvious what to do here, do we still want to deprecate this? Should we talk about it in the call tomorrow?

**Comment by SkySkimmer** [comment]
What's non obvious? We do want to deprecate it.

**Comment by jfehrle** [comment]
Deprecation now is fine, but we may want to discuss configuration in general before we actually remove it in the future.  And sometimes we change our minds on deprecation.  

FWIW, I'd like to eventually post some use cases with which we can evaluate how adequate current configuration mechanisms are and perhaps identify possible improvements.

**Comment by Alizter** [comment]
@SkySkimmer In which case can you add Enrico's suggestion (or reject it if you want).

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
Please apply [Enrico's suggestion](https://github.com/coq/coq/pull/15652#discussion_r803670441) (and also add a similar description to the deprecation notice in the manual).

**Comment by SkySkimmer** [comment]
updated

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
will merge later today (and when the change request is lifted)

**Review by Zimmi48 — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**